src/core/components/pin/pin-manager.js

```diff
@@ -106,43 +107,65 @@ class PinManager {
   // Encode and write pin key sets to the datastore:
   // a DAGLink for each of the recursive and direct pinsets
   // a DAGNode holding those as DAGLinks, a kind of root pin
-  async flushPins () {
-    const [
-      dLink,
-      rLink
-    ] = await Promise.all([
-      // create a DAGLink to the node with direct pins
-      this.pinset.storeSet(this.directKeys())
-        .then((result) => {
-          return new DAGLink(PinTypes.direct, result.node.size, result.cid)
-        }),
-      // create a DAGLink to the node with recursive pins
-      this.pinset.storeSet(this.recursiveKeys())
-        .then((result) => {
-          return new DAGLink(PinTypes.recursive, result.node.size, result.cid)
-        }),
-      // the pin-set nodes link to a special 'empty' node, so make sure it exists
-      this.dag.put(new DAGNode(Buffer.alloc(0)), {
-        version: 0,
-        format: multicodec.DAG_PB,
-        hashAlg: multicodec.SHA2_256,
-        preload: false
-      })
-    ])
+  async flushPins () { // eslint-disable-line require-await
+    if (this._flushingPins) {
+      this._flushingPins.controller.abort()
+    }
 
-    // create a root node with DAGLinks to the direct and recursive DAGs
-    const rootNode = new DAGNode(Buffer.alloc(0), [dLink, rLink])
-    const rootCid = await this.dag.put(rootNode, {
-      version: 0,
-      format: multicodec.DAG_PB,
-      hashAlg: multicodec.SHA2_256,
-      preload: false
-    })
+    const controller = new AbortController()
+
+    this._flushingPins = {
+      controller,
+      promise: delay(100)
+        .then(async () => {
```

Comment:
How come this call has a `.then().catch()`? We should stick to `await` as much as possible.

```js
promise: (async () => {
  await delay(100)
  if (controller.signal.aborted) return
  // ...
})()
```

Reply:
Personal preference I guess?  I've always found IIFEs a bit, well iffy to look at.